flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1955,6 +1963,25 @@ private void abortPendingCheckpoint(
         }
     }
 
+    private void handleCheckpointException(
+            PendingCheckpoint pendingCheckpoint,
+            CheckpointProperties checkpointProperties,
+            CheckpointException exception,
+            ExecutionAttemptID executionAttemptID) {
+        if (checkpointProperties.isSavepoint()
+                && (checkpointProperties.isSynchronous() || pendingCheckpoint == null)) {
+            failureManager.handleSynchronousSavepointFailure(exception);
+        } else if (executionAttemptID != null && pendingCheckpoint != null) {
+            failureManager.handleTaskLevelCheckpointException(
+                    exception, pendingCheckpoint.getCheckpointId(), executionAttemptID);
+        } else if (pendingCheckpoint != null) {
+            failureManager.handleJobLevelCheckpointException(
+                    exception, pendingCheckpoint.getCheckpointId());
+        } else {
+            failureManager.handleJobLevelCheckpointException(exception);
+        }
+    }
```

Comment:
Wouldn't it be better to have this logic inside `CheckpointFailureManager`? On this level it's pretty obfuscated what differs between `handleSynchronousSavepointFailure` `handleJobLevelCheckpointException` anyway.